go/logger/buffer.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2019 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package logger
+
+import (
+	"bufio"
+	"io"
+	"sync"
+	"time"
+)
+
+type autoFlushingBufferedWriter struct {
+	lock           sync.RWMutex
+	bufferedWriter *bufio.Writer
+	backupWriter   *bufio.Writer
+
+	frequency        time.Duration
+	timer            *time.Timer
+	currentlyWaiting bool
+	shutdown         chan struct{}
+}
+
+var _ io.Writer = &autoFlushingBufferedWriter{}
+
+func (writer *autoFlushingBufferedWriter) backgroundFlush() {
+	for {
+		select {
+		case <-writer.timer.C:
+			// Swap out active and backup writers
+			writer.lock.Lock()
+			writer.currentlyWaiting = false
+			previouslyActive := writer.bufferedWriter
+			writer.bufferedWriter = writer.backupWriter
+			writer.lock.Unlock()
+
+			// Flush the previously active writer and complete the swap. We do
+			// not need to complete the swap under lock because this is the
+			// only goroutine accessing backupWriter.
```

Comment:
Technically true, but IMO following my suggestion above results in better code clarity.